src/spec/expectations.cr

```diff
@@ -5,17 +5,49 @@ module Spec
     end
 
     def match(actual_value)
-      actual_value == @expected_value
+      expected_value = @expected_value
+
+      # For the case of comparing strings we want to make sure that two strings
+      # are equal if their content is equal, but also their bytesize and size
+      # should be equal. Otherwise, an incorrect bytesize or size was used
+      # when creating them.
+      if actual_value.is_a?(String) && expected_value.is_a?(String)
+        actual_value == expected_value &&
+          actual_value.bytesize == expected_value.bytesize &&
+          actual_value.size == expected_value.size
+      else
+        actual_value == @expected_value
+      end
     end
 
     def failure_message(actual_value)
-      expected = @expected_value.inspect
-      got = actual_value.inspect
-      if expected == got
-        expected += " : #{@expected_value.class}"
-        got += " : #{actual_value.class}"
+      expected_value = @expected_value
+
+      # Check for the case of string equality when the content match
+      # but not the bytesize or size.
```

Comment:
How about recalculating the size of both the actual and expected values and report an error that either the actual or expected values are invalid? This would be more helpful and precise information, and it would be easy to calculate.

Reply:
Sorry, I don't understand what you mean.

Reply:
instead of just comparing the bytesizes, work out their actual bytesize using `Char::Reader` and use that as the reference

Reply:
I know, you are talking about the case when you pass a computed string to `eq`. I think doing that is a bit overkill.

Reply:
Sorry for the long conversation, but for example one might do:

```crystal
some_computed_string = ...
some_computed_string.should eq("foo" * 3)
```

Now, there might be a bug in `"foo" * 3` that returns an invalid `bytesize` or `size`. But `"foo" * 3` is already tested in the standard library. If you use another custom method as an argument to `eq`, you should test it separately, and of course you won't test it by supplying values that use the same function that you are testing... so I think assuming what you pass in `eq` already has correct values of `bytesize` and `size` is fine.